### PR TITLE
[ Awaiting Review ] Fix advection limitation, improve 'How It works' page

### LINKED ISSUE #82: Plotting example for Nonlinear problems
I tried to simulate a 1D wave on a string with fixed ends and plotted the resulting animation and I'm posing the code here so that it could be refined into a tutorial.
The code is:
```julia
using ModelingToolkit
using MethodOfLines
using DomainSets
using NonlinearSolve
using Plots
# using SciMLNLSolve

@parameters t x
@variables E(..) u(..)

∂t = Differential(t)
∂tt = ∂t^2
∂xx = Differential(x)^2

c = 1

wave_eq1D = ∂tt(u(x, t)) ~ c^2 * ∂xx(u(x, t))

f(x) = sin(x)
g(x) = 0

bcs = [
    u(0, t) ~ 0,
    u(3, t) ~ 0,
    u(x, 0) ~ f(x),
    ∂t(u(x, 0)) ~ g(x)
]

t_min = 0.0
t_max = 2.0
x_min = 0.0
x_max = 3.0

# Space and time domains
domains = [
    t ∈ Interval(t_min, t_max),
    x ∈ Interval(x_min, x_max)
]

@named pde_system = PDESystem(wave_eq1D, bcs, domains, [t, x], [u(x, t)])

# Method of lines discretization
dx = 0.03
dt = 0.04

Nx = floor(Int64, (x_max - x_min) / dx) + 1

order = 2
discretization = MOLFiniteDifference([x => dx, t => dt], approx_order=2)

# Convert the PDE problem into an nonlinear problem
prob = discretize(pde_system, discretization)

sol = solve(prob, NewtonRaphson(), tol=1e-5)

anim = @animate for i in eachindex(t_min:dt:t_max)
    scatter(x_min:dx:x_max, sol[(i-1)*Nx+1:i*Nx], y_lims=(-1,1))
end

gif(anim, "wave.gif", fps=15)
```
wave

The code needs a bit of cleanup and I'll also have to check if the chosen `dt` and `dx` make sense given the stability criteria. Moreover, the problem is a bit big for an example / tutorial since the string length is quite big with respect to the natural wavelength of the problem. I'll try to find more appropriate parameters over the following days.

Thanks to @xtalax for helping me with getting the above to work.

### LINKED ISSUE #85: Solving nonlinear PDE
I am trying to solve this nonlinear PDE du/dt + u du/dx = .....  but gives error that u is not defined. I am posting the code and the error here.
```
using ModelingToolkit, MethodOfLines, DomainSets, NonlinearSolve, DifferentialEquations

# Define the system of equations
@parameters t x 
@variables u(..) 

Dt = Differential(t)
Dx = Differential(x)


tmin = 0.;
tmax = 6000.;

xmin = 0.;
xmax = 200.;


dx = 10.0;
dt = 3.0;
order = 2; # Order of the finite element approximation, must be even



eqs = [#Dt(u(t,x)) + u(t,x) * Dx(u(t,x)) ~ 0.0
        Dt(u(t,x)) + 0.5 * Dx(u(t,x)^2) ~ 0.0]    

## Note that we are defining hu(t,x) because we cannot do 
domain = [x ∈ Interval(xmin,xmax), 
          t ∈ Interval(tmin,tmax)]

# Define the BC function 

bcs = [u(tmin, x) ~ 0.,
       u(t, xmin) ~ 0.,
       Dt(u(t, xmax))+ 3 *  Dx(u(t,xmax)) ~ 0.0]


@named pdesys = PDESystem(eqs, bcs, domain, [t, x], [u(t,x)])

# Create discretization


discretization = MOLFiniteDifference([x => dx], t, approx_order = order, grid_align = center_align) # edge_align gives better results with Neumann BCs, though uses extra grid points

prob = ModelingToolkit.discretize(pdesys, discretization)

sol = NonlinearSolve.solve(prob, Tsit5())

```

### LINKED ISSUE #100: Structural simplify Nonlinear problems
Due to recent upgrades to MTK, it is now possible to do this for performance and retrieve the solution. Credit to @YingboMa.

### PR DESCRIPTION
**Pull request by xtalax**
Following from discussion [here](https://discourse.julialang.org/t/dirac-delta-in-boundary-conditions-for-methodoflines-jl/80567/8) and on slack

Adds Steady state tutorial and improves docs

fixes some broken tests by updating to use `get_discrete`

fixes #85 
fixes #82 
fixes #100

### CODE REVIEW

File: docs/src/index.md [hunk 1]
```diff
@@ -34,9 +34,11 @@ At the moment the package is able to discretize almost any system, with some ass
 
 - That the grid is cartesian.
 - That the equation is first order in time.
+- Boundary conditions in time are supplied as initial conditions, not at the end of the simulation interal. If your system requires a final condition, please use a change of variables to rectify this.
```
Comment (ChrisRackauckas) [inline docs/src/index.md:37]: maybe not that this is just a current limitation?

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by xtalax** [comment]
@ChrisRackauckas why is the docbuild failing?

**Comment by ChrisRackauckas** [comment]
Add ModelingToolkit as a module for it.

**Comment by xtalax** [comment]
@ChrisRackauckas Ok added deps and fixed. This new error is hard to diagnose

**Comment by xtalax** [comment]
I don't think I am showing one anywhere, which is why this confused me. What needs to go in `systems`?

**Comment by ChrisRackauckas** [comment]
It'll call show automatically on any `PDESystem` at the bottom of a block.

**Comment by xtalax** [comment]
There is a segfault in the bruss example on the CI, but this is not present locally

File: docs/src/tutorials/params.md [hunk 2]
```diff
@@ -69,15 +80,16 @@ for (Dnval, Dpval) in zip(rand(10), rand(10))
     push(sols, solve(newprob, Tsit5()))
 end
 
-x = (0:0.1:1)[2:end-1]
-t = sol.t
+grid = get_discrete(pdesys, discretization)
+discrete_x = grid[x]
+discrete_t = sol[t]
```
Comment (MilkshakeForReal) [inline docs/src/tutorials/params.md:85]: It should be 
`discrete_t = sols[t]`

File: docs/src/tutorials/params.md [hunk 3]
```diff
@@ -41,43 +48,48 @@ gif(anim, "plot.gif",fps=30)
...
 
-pdesys = PDESystem(eqs,bcs,domains,[t,x],[u(t,x),v(t,x)],[Dn=>0.5, Dp=>2])
-discretization = MOLFiniteDifference([x=>0.1],t)
 prob = discretize(pdesys,discretization) # This gives an ODEProblem since it's time-dependent
 
 sols = []
 for (Dnval, Dpval) in zip(rand(10), rand(10))
-    newprob = remake(prob, p = [Dn => Dnval, Dp => Dpval])
+    newprob = remake(prob, ps = [Dnval, Dpval])
     push(sols, solve(newprob, Tsit5()))
```
Comment (MilkshakeForReal) [inline docs/src/tutorials/params.md:80]: it should be `push!`

Comment (xtalax) [reply docs/src/tutorials/params.md]: Thanks!

**Comment by ChrisRackauckas** [comment]
@xtalax what's left here?

**Comment by xtalax** [comment]
@ChrisRackauckas Weird docs segfault, looks to be do to running the examples, fixing a stale test that wasn't being run.

Could I move problem doc tests to the interface tests to check that they run, to avoid the segfault? A lot of them are already in there.

**Comment by ChrisRackauckas** [comment]
Is it segfaulting in the `show`? What about making sure to `show` different output?

**Comment by xtalax** [comment]
As for the last segfault, the error provided literally no context, not even a file

**Comment by ChrisRackauckas** [comment]
Awesome, big step forward.
